tfmigrate/migrator.go

```diff
@@ -21,36 +23,42 @@ type Migrator interface {
 }
 
 // setupWorkDir is a common helper function to setup work dir and returns the
-// current state and a swtich back function.
-func setupWorkDir(ctx context.Context, tf tfexec.TerraformCLI) (*tfexec.State, func(), error) {
+// current state and a switch back function.
+func setupWorkDir(ctx context.Context, tf tfexec.TerraformCLI, workspace string) (*tfexec.State, func(), error) {
 	// check if terraform command is available.
 	version, err := tf.Version(ctx)
 	if err != nil {
 		return nil, nil, err
 	}
 	log.Printf("[INFO] [migrator@%s] terraform version: %s\n", tf.Dir(), version)
 
-	// initialize work dir.
+	// create local workspace folder
+	path := filepath.Join(tf.Dir(), "terraform.tfstate.d", workspace)
+	log.Printf("[INFO] [migrator@%s] creating local workspace folder in: %s\n", tf.Dir(), path)
+	os.MkdirAll(path, os.ModePerm)
```

Comment:
Because the clean up logic for the `terraform.tfstate.d` is implemented in the defer function of `OverrideBackendToLocal`, I guess it will fail to clean up if the `tf.Init`, `tf.StatePull` or `tf.WorkspaceSelect` returns an error.

If my understanding is correct, the `terraform.tfstate.d` is only for the local backend and it doesn't need for remote backends. If so, can we move creating the dir into `OverrideBackendToLocal`?

Reply:
that's correct. the only requirement is that the `terraform.tfstate.d` folder exist when doing `terraform init -reconfigure` so moving it to the `OverrideBackendToLocal` folder also works.